apollo-router/src/plugins/telemetry/tracing/otlp.rs

```diff
@@ -0,0 +1,311 @@
+use crate::configuration::ConfigurationError;
+use crate::plugins::telemetry::config::Trace;
+use crate::plugins::telemetry::{GenericWith, TracingConfigurator};
+#[cfg(feature = "otlp-grpc")]
+use grpc::*;
+use opentelemetry::sdk::trace::Builder;
+use opentelemetry_otlp::{SpanExporterBuilder, WithExportConfig};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::result::Result;
+use std::time::Duration;
+use tower::BoxError;
+use url::Url;
+
+#[cfg(feature = "otlp-http")]
+use self::http::*;
+
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
+#[serde(deny_unknown_fields)]
+pub struct Config {
+    pub endpoint: Endpoint,
+    pub protocol: Option<Protocol>,
+
+    #[serde(deserialize_with = "humantime_serde::deserialize")]
+    #[schemars(with = "String")]
+    pub timeout: Option<Duration>,
+    #[cfg(feature = "otlp-grpc")]
+    pub grpc: Option<GrpcExporter>,
+    #[cfg(feature = "otlp-http")]
+    pub http: Option<HttpExporter>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
+#[serde(deny_unknown_fields, rename_all = "snake_case", untagged)]
+pub enum Endpoint {
+    Default(EndpointDefault),
+    Url(Url),
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
+pub enum EndpointDefault {
+    Default,
+}
```

Comment:
Could also be replaced with the shared enum `AgentEndpoint` if I'm not wrong

Reply:
There is a difference between these.
Datadog and Otlp endpoint accept a URL
Zipkin and Jaeger accept a SocketAddress.

If they were all the same I would have gone ahead and refactored, but as there are two flavors it's less compelling.

I'm tempted to say we should keep these separate in case socket address also becomes an option for OTLP. At least each config is fully contained in the file it is in.

Reply:
Fixed in [9a507cc](https://github.com/apollographql/router/pull/782/commits/9a507cc6247153d4ef407654c586c145a7df5b3e)